src/compiler/checker.ts

```diff
@@ -14095,7 +14104,7 @@ namespace ts {
                 }
             }
 
-            return getUnionType(map(signatures, ctor ? getJsxPropsTypeFromConstructSignature : getJsxPropsTypeFromCallSignature), UnionReduction.None);
+            return getUnionType(map(signatures, ctor ? isJs ? getJsxPropsTypeFromConstructSignatureJs : getJsxPropsTypeFromConstructSignature : getJsxPropsTypeFromCallSignature), UnionReduction.None);

```

Comment:
It seems a lot easier to read to write this as `x => getJsxPropsTypeFromConstructSignature(x, isJs)` — and everywhere else you added function pairs.

Reply:
But that'd introduce a lambda on every invocation here, and this is called in contextual typing (which is done often), whereas this way only two functions get made ever. I do agree that that looks better, I just think it'd be detrimental to perf. :(

Reply:
[1] Did you measure? You’re still doing dynamic dispatch here so I can’t imagine the numbers would be that different (though I am not sure how good the VMs are at lifting lambdas).
[2] At least extract the double-nested conditional to a local. That’s just egregious. :]